README.md
# nordvpn-indicator
Linux system tray icon for nordvpn
This will configure the nordvpn repository and install the nordvpn package on first boot.

## Translations
You can help translating nordvpn-indicator on [Transifex](https://www.transifex.com/abalfoort/nordvpn-indicator)



debian/changelog
nordvpn-indicator (1.9.8) stable; urgency=low

  * Updated translation: pt

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Tue, 27 Dec 2022 16:57:14 +0100

nordvpn-indicator (1.9.7) stable; urgency=low

  * Updated translation: es

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Fri, 07 Oct 2022 19:13:35 +0200

nordvpn-indicator (1.9.6) stable; urgency=low

  * Updated translation: es_419 (latin america)

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Tue, 28 Jun 2022 08:31:44 +0200

nordvpn-indicator (1.9.5) stable; urgency=low

  * Updated translation: ca

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Wed, 14 Apr 2021 08:38:01 +0200

nordvpn-indicator (1.9.4) stable; urgency=low

  * Add dependency: gir1.2-notify-0.7

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Wed, 20 Jan 2021 12:04:23 +0100

nordvpn-indicator (1.9.3) stable; urgency=low

  * Add dependency: | gir1.2-ayatanaappindicator3-0.1

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Thu, 31 Dec 2020 17:08:16 +0100

nordvpn-indicator (1.9.2) stable; urgency=low

  * Updated translation: it

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Mon, 02 Nov 2020 11:40:45 +0100

nordvpn-indicator (1.9.1) stable; urgency=low

  * Install nordvpn through nordvpn repository

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Thu, 22 Oct 2020 09:50:20 +0200

nordvpn-indicator (1.9.0) stable; urgency=low

  * Let nordvpn handle quick connect
  * Select servers that have a load > 0

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Mon, 19 Oct 2020 23:10:45 +0200

nordvpn-indicator (1.8.9) stable; urgency=low

  * Move data files installation from setup.py to nordvpn-indicator.ini
  * Move nordvpn-indicator.conf to .config/nordvpn/inidcator.conf

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Tue, 29 Sep 2020 10:10:32 +0200

nordvpn-indicator (1.8.8) stable; urgency=low

  * Convert to Debian Python packaging

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Wed, 23 Sep 2020 11:40:51 +0200

nordvpn-indicator (1.8.7) stable; urgency=low

  * Updated translation: da

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Wed, 17 Jun 2020 09:00:23 +0200

nordvpn-indicator (1.8.6) stable; urgency=low

  * Fix wrong status when not connected to the internet

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Tue, 05 May 2020 18:10:09 +0200

nordvpn-indicator (1.8.5) stable; urgency=low

  * Fix icon always connected
  * Put default order URL in nordvpn-indicator.conf

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Sun, 19 Apr 2020 07:57:37 +0200

nordvpn-indicator (1.8.4) stable; urgency=low

  * Updated translation: de

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Wed, 26 Feb 2020 08:18:32 +0100

nordvpn-indicator (1.8.3) stable; urgency=low

  * Updated translation: lt

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Thu, 30 Jan 2020 21:14:32 +0100

nordvpn-indicator (1.8.2) stable; urgency=low

  * Show connection rating result
  * Update deprecated code
  * Updated translation: pt

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Mon, 13 Jan 2020 15:41:15 +0100

nordvpn-indicator (1.8.1) stable; urgency=low

  * Add connection rating

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Tue, 07 Jan 2020 11:15:31 +0100

nordvpn-indicator (1.8.0) stable; urgency=low

  * Fix bug showing not connected on slow VPN

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Fri, 06 Dec 2019 11:24:20 +0100

nordvpn-indicator (1.7.9) stable; urgency=low

  * Updated translation: pt

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Sun, 10 Nov 2019 10:46:47 +0100

nordvpn-indicator (1.7.8) stable; urgency=low

  * Updated translation: da

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Fri, 08 Nov 2019 17:26:48 +0100

nordvpn-indicator (1.7.7) stable; urgency=low

  * Fix country names with spaces not listed

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Tue, 08 Oct 2019 15:39:03 +0200

nordvpn-indicator (1.7.6) stable; urgency=low

  * Resolve bug not able to connect through menu

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Sun, 15 Sep 2019 16:57:41 +0200

nordvpn-indicator (1.7.5) stable; urgency=low

  * Check for daemon.log to decide account availability

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Tue, 10 Sep 2019 13:34:34 +0200

nordvpn-indicator (1.7.4) stable; urgency=low

  * Updated translation: ca

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Sat, 31 Aug 2019 21:26:13 +0200

nordvpn-indicator (1.7.3) stable; urgency=low

  * Better format status information

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Fri, 30 Aug 2019 11:59:40 +0200

nordvpn-indicator (1.7.2) stable; urgency=low

  * Updated translations: de, li

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Thu, 29 Aug 2019 17:01:23 +0200

nordvpn-indicator (1.7.1) stable; urgency=low

  * Updated translation: ca

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Sun, 25 Aug 2019 08:24:01 +0200

nordvpn-indicator (1.7.0) stable; urgency=low

  * Change xdg autostart variables

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Fri, 23 Aug 2019 13:37:57 +0200

nordvpn-indicator (1.6.9) stable; urgency=low

  * Move remaining nordvpn commands to own class

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Tue, 20 Aug 2019 20:08:36 +0200

nordvpn-indicator (1.6.8) stable; urgency=low

  * Add manual connecting

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Wed, 14 Aug 2019 13:29:00 +0200

nordvpn-indicator (1.6.7) stable; urgency=low

  * Incorporate NordVPN order link
  * Fix Gtk+ deprecated functions

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Tue, 13 Aug 2019 19:50:41 +0200

nordvpn-indicator (1.6.6) stable; urgency=low

  * Let user select NordLynx if possible

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Sun, 11 Aug 2019 13:43:01 +0200

nordvpn-indicator (1.6.5) stable; urgency=low

  * Check fastest server if manually connecting
  * Implement changes for use with WireGuard

 -- Arjen Balfoort <arjenbalfoort@hotmail.com>  Fri, 09 Aug 2019 11:45:58 +0200


nordvpn-indicator/__init__.py
#! /usr/bin/env python3

"""
Dependencies: nordvpn, gir1.2-gtk-3.0, gir1.2-appindicator3-0.1
This script assumes successful login
Auto login with: nordvpn set autoconnect enabled [COUNTRY]
Get a list of countries with: nordvpn countries
"""

APPINDICATOR_ID = 'nordvpn-indicator'
INTERVAL = 10

import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk
try:
    gi.require_version('AppIndicator3', '0.1')
    from gi.repository import AppIndicator3
except:
    gi.require_version('AyatanaAppIndicator3', '0.1')
    from gi.repository import AyatanaAppIndicator3 as AppIndicator3
gi.require_version('Notify', '0.7')

import subprocess
import signal
from threading import Event, Thread
from gi.repository import Notify
from os.path import abspath, dirname, join, exists
from pathlib import Path
import re
from os import makedirs

# Local modules
from .login import NordVPNLogin
from .connect import NordVPNConnect
from .settings import NordVPNSettings
from .nordvpn import is_loggedin, is_connected, get_fastest_server, \
                    has_account, load_order_page, get_account_info, \
                    get_status_info, get_connection_status, \
                    nordvpn_connect, nordvpn_disconnect, rate_connection

# i18n: http://docs.python.org/3/library/gettext.html
import gettext
_ = gettext.translation(APPINDICATOR_ID, fallback=True).gettext


class NordVPNIndicator():
    def __init__(self):
        """
        Provides tray icon with menu and checks
        every INTERVAL for changes in the connection.
        """
        # Save current directory
        self.script_dir = abspath(dirname(__file__))
        # Translations (used in multiple functions)
        self.order_text = _('Get a NordVPN account')
        self.connections = {
            'connecting': {'label': _('Quick connect'), 'icon': join(self.script_dir, 'connecting.svg')},
            'disconnecting': {'label': _('Quick connect'), 'icon': join(self.script_dir, 'connecting.svg')},
            'connected': {'label': _('Disconnect'), 'icon': join(self.script_dir, 'connected.svg')},
            'disconnected': {'label': _('Quick connect'), 'icon': join(self.script_dir, 'disconnected.svg')},
            'no_internet': {'label': _('Quick connect'), 'icon': join(self.script_dir, 'connecting.svg')}
        }
        self.manual_connect_text = _('Manual connect')
        self.status_text = _('Status Information')
        self.rate_text = _('Rate last connection')
        self.poor_text = _('poor')
        self.excellent_text = _('excellent')
        self.loggedin_text = _('You are not logged into NordVPN.\n'
                               'Please, login with: nordvpn login')
        
        self.current_connection = 'connecting'
        # Current NordVPN settings
        self.current_settings = {}
        self.fill_settings(True)
        self.settings_changed = False
        # Create event to use when thread is done
        self.check_done_event = Event()
        # Create indicator object
        self.indicator = AppIndicator3.Indicator.new(APPINDICATOR_ID, self.connections[self.current_connection]['icon'], AppIndicator3.IndicatorCategory.SYSTEM_SERVICES)
        self.indicator.set_title('NordVPN Indicator')
        self.indicator.set_status(AppIndicator3.IndicatorStatus.ACTIVE)
        self.indicator.set_menu(self.build_menu())
        # Init notifier
        Notify.init(APPINDICATOR_ID)
        # Start thread to check for connection changes
        Thread(target=self.run_check).start()
        # Some debugging
        print(('--------- NordVPNIndicator Init ---------'))

    def fill_settings(self, force=False):
        """
        Get the current NordVPN settings in a dictionary
        You need to be logged in to get these settings
        """
        if force or self.settings_changed:
            settings = {'protocol': '', 'country': '', 'server': ''}
            try:
                output = subprocess.check_output(['nordvpn', 'settings']).decode('utf-8').lower().strip()
            except:
                output = ''
            if 'dns' in output:
                output = output.replace(' ', '').replace('-', '').replace('_', '')
                output_list = output.split('\n')
                for line in output_list:
                    # extract settings name and value
                    # DeprecationWarning: invalid escape sequence \w > mark the regex as raw string (prefix r)
                    match = re.search(r'([\w]+):([\w]+)', line)
                    if match:
                        value = match.group(2)
                        if 'enabled' in value: value = True
                        elif 'disabled' in value: value = False
                        settings[match.group(1)] = value
                # Add country details (because 'nordvpn settings' doesn't show it)
                nordvpn_save = '{}/.config/nordvpn'.format(str(Path.home()))
                if not exists(nordvpn_save): makedirs(nordvpn_save)
                settings['nordvpnsave'] = nordvpn_save
                if settings['autoconnect']:
                    server_save = join(nordvpn_save, 'server')
                    country_save = join(nordvpn_save, 'country')
                    if exists(server_save):
                        with open(server_save, 'r') as f:
                            settings['server'] = f.read().replace('\n', '').replace('\r', '')
                    elif exists(country_save):
                        with open(country_save, 'r') as f:
                            settings['country'] = f.read().replace('\n', '').replace('\r', '')

            # Save in the current_settings dictionary
            self.current_settings = settings
        # Show current settings
        print(('---------------- Settings ---------------'))
        print(("{}".format(', '.join("{!s}={!r}".format(key,val) for (key,val) in self.current_settings.items()))))

    def build_menu(self):
        """
        Build menu for the tray icon
        """
        menu = Gtk.Menu()
        if not has_account():
            self.item_order = Gtk.MenuItem.new_with_label(self.order_text)
            self.item_order.connect('activate', self.show_order_page)
            menu.append(self.item_order)
            menu.append(Gtk.SeparatorMenuItem())
        item_quick_connect = Gtk.MenuItem.new_with_label(self.connections[self.current_connection]['label'])
        item_quick_connect.connect('activate', self.quick_connect)
        menu.append(item_quick_connect)
        item_manual_connect = Gtk.MenuItem.new_with_label(self.manual_connect_text)
        item_manual_connect.connect('activate', self.manual_connect)
        menu.append(item_manual_connect)
        
        # Rating
        item_rate = Gtk.MenuItem.new_with_label(self.rate_text)
        sub_menu = Gtk.Menu()
        sub_item_rate_1 = Gtk.MenuItem.new_with_label('1 ({})'.format(self.poor_text))
        sub_item_rate_1.connect('activate', self.rate_prev_connection, 1)
        sub_menu.append(sub_item_rate_1)
        sub_item_rate_2 = Gtk.MenuItem.new_with_label('2')
        sub_item_rate_2.connect('activate', self.rate_prev_connection, 2)
        sub_menu.append(sub_item_rate_2)
        sub_item_rate_3 = Gtk.MenuItem.new_with_label('3')
        sub_item_rate_3.connect('activate', self.rate_prev_connection, 3)
        sub_menu.append(sub_item_rate_3)
        sub_item_rate_4 = Gtk.MenuItem.new_with_label('4')
        sub_item_rate_4.connect('activate', self.rate_prev_connection, 4)
        sub_menu.append(sub_item_rate_4)
        sub_item_rate_5 = Gtk.MenuItem.new_with_label('5 ({})'.format(self.excellent_text))
        sub_item_rate_5.connect('activate', self.rate_prev_connection, 5)
        sub_menu.append(sub_item_rate_5)
        item_rate.set_submenu(sub_menu)
        menu.append(item_rate)
        
        item_status = Gtk.MenuItem.new_with_label(self.status_text)
        item_status.connect('activate', self.show_status)
        menu.append(item_status)
        item_settings = Gtk.MenuItem.new_with_label(_('Settings'))
        item_settings.connect('activate', self.show_settings)
        menu.append(item_settings)
        menu.append(Gtk.SeparatorMenuItem())
        item_quit = Gtk.MenuItem.new_with_label(_('Quit'))
        item_quit.connect('activate', self.quit)
        menu.append(item_quit)
        menu.show_all()
        
        if self.current_connection == 'connecting' or \
           self.current_connection == 'disconnecting' or \
           self.current_connection == 'no_internet':
            # Connecting/disconnecting/no internet
            item_quick_connect.set_sensitive(False)
            item_manual_connect.set_sensitive(False)
            item_settings.set_sensitive(False)
            item_status.set_sensitive(False)
            item_rate.set_sensitive(False)
        elif self.current_connection == 'connected':
            # Connected
            item_quick_connect.set_sensitive(True)
            item_manual_connect.set_sensitive(False)
            item_settings.set_sensitive(True)
            item_status.set_sensitive(True)
            item_rate.set_sensitive(False)
        else:
            # Disconnected
            item_quick_connect.set_sensitive(True)
            item_manual_connect.set_sensitive(True)
            item_settings.set_sensitive(False)
            item_status.set_sensitive(True)
            item_rate.set_sensitive(True)
                
        return menu

    def run_check(self):
        """
        Checks every INTERVAL for changes in the connection.
        """
        while not self.check_done_event.is_set():
            connection = get_connection_status()
            if connection != self.current_connection:
                self.current_connection = connection
                print(('Connection status: {}'.format(self.current_connection)))
                # Change icon
                self.indicator.set_icon_full(self.connections[self.current_connection]['icon'], '')
                # Build menu
                self.indicator.set_menu(self.build_menu())
            self.check_done_event.wait(INTERVAL)

    def rate_prev_connection(self, widget, rate):
        """
        Rate the last connection
        """
        return_code, rate_result = rate_connection(rate)
        icon = 'dialog-ok'
        if return_code > 0:
            icon = 'dialog-error'
        # Show rate info in notification window
        Notify.Notification.new(self.rate_text, rate_result, icon).show()
    
    def show_status(self, widget=None):
        """
        Show a notification with status information
        of the current connection.
        """
        if is_connected():
            email, expires = get_account_info()
            # Use box horizontal character (dec 9472)
            text = '{0}\n{1}\n{2}'.format(expires, chr(9472) * 25, get_status_info())
            print(('-------------- show_status --------------'))
            print(text)
            icon = 'dialog-warning' if 'Disconnected' in text else 'dialog-information'
        else:
            text = self.loggedin_text
            icon = 'dialog-error'
        # Show status info in notification window
        Notify.Notification.new(self.status_text, text, icon).show()

    def show_settings(self, widget):
        """
        Show the settings window.
        """
        self.settings_changed = NordVPNSettings(self.current_settings).show_settings()
        self.fill_settings()
    
    def quick_connect(self, widget):
        """
        Quick connect.
        """
        self.change_connection(quick=True)
    
    def manual_connect(self, widget):
        """
        Show the connect window.
        """
        country, server = NordVPNConnect().show_connect()
        if country or server:
            self.change_connection(country=country, server=server)
        
    def show_order_page(self, widget=None):
        """
        Load the order page
        """
        load_order_page()

    def change_connection(self, country=None, server=None, connect=None, quick=False):
        """
        Start connection change in a thread
        Login when needed
        """
        return_code = 0
        if not is_loggedin():
            # Show login window
            return_code, last_line = NordVPNLogin().show()

        if return_code == 0:
            # Start this threaded or else the system tray icon does not change
            Thread(target=self.run_change_connection, kwargs={'country': country,'server': server, 'connect': connect, 'quick': quick}).start()
        else:
            # Failed to login
            title = _('Not logged into NordVPN.')
            Notify.Notification.new(title, last_line, 'dialog-error').show()
            print(('----------- change_connection -----------'))
            print(last_line)
            
    def run_change_connection(self, country=None, server=None, connect=None, quick=False):
        """
        Switches connection:
        Disconnect when connected and vise versa.
        """
        # Save current connection status
        if connect is None: connect = not is_connected()

        # Connect to country/server or disconnect
        connect_obj = ''
        if not quick and connect:
            if not country:
                country = self.current_settings['country']
            if not server:
                server = self.current_settings['server']
            connect_obj = server if server else country
            if not connect_obj:
                # Get fastest server to connect to
                connect_obj = get_fastest_server()

        if connect:
            return_code, output = nordvpn_connect(connect_obj)
        else:
            return_code, output = nordvpn_disconnect()
        # Show notification of error
        if return_code != 0:
            if not output:
                output = _('If the problem persists, contact NordVPN customer support.')
            error_title = _('Failed to connect to "{0}"'.format(connect_obj)) if connect else _('Failed to disconnect from "{0}"'.format(connect_obj))
            Notify.Notification.new(error_title, output, 'dialog-error').show()

    def quit(self, widget=None):
        """
        Quit the application.
        """
        self.check_done_event.set()
        Notify.uninit()
        Gtk.main_quit()

def main():
    NordVPNIndicator()
    signal.signal(signal.SIGINT, signal.SIG_DFL)
    Gtk.main()
    
if __name__ == '__main__':
    main()
